File: src/go/3_StructuralPatterns/3_Composite/neuralnetworks.go
package composite

type NeuronInterface interface {
  Iter() []*Neuron
}

type Neuron struct {
  In, Out []*Neuron
}

func (n *Neuron) Iter() []*Neuron {
  return []*Neuron{n}
}

func (n *Neuron) ConnectTo(other *Neuron) {
  n.Out = append(n.Out, other)
  other.In = append(other.In, n)
}

type NeuronLayer struct {
  Neurons []Neuron
}

func (n *NeuronLayer) Iter() []*Neuron {
  result := make([]*Neuron, 0)
  for i := range n.Neurons {
    result = append(result, &n.Neurons[i])
  }
  return result
}

func NewNeuronLayer(count int) *NeuronLayer {
  return &NeuronLayer{make([]Neuron, count)}
}

func Connect(left, right NeuronInterface) {
  for _, l := range left.Iter() {
    for _, r := range right.Iter() {
      l.ConnectTo(r)
    }
  }
}

func main() {
  neuron1, neuron2 := &Neuron{}, &Neuron{}
  layer1, layer2 := NewNeuronLayer(3), NewNeuronLayer(4)

  //neuron1.ConnectTo(&neuron2)

  Connect(neuron1, neuron2)
  Connect(neuron1, layer1)
  Connect(layer2, neuron1)
  Connect(layer1, layer2)
}

File: src/csharp/3_StructuralPatterns/3_Composite/NeuralNetworks.cs
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace DotNetDesignPatternDemos.Structural.Composite.NeuralNetworks
{
  // we cannot use a base class 

  public static class ExtensionMethods
  {
    public static void ConnectTo(this IEnumerable<Neuron> self, IEnumerable<Neuron> other)
    {
      if (ReferenceEquals(self, other)) return;

      foreach (var from in self)
        foreach (var to in other)
        {
          from.Out.Add(to);
          to.In.Add(from);
        }
    }
  }

  public class Neuron : IEnumerable<Neuron>
  {
    public float Value;
    public List<Neuron> In, Out;

    public IEnumerator<Neuron> GetEnumerator()
    {
      yield return this;
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
      yield return this;
    }
  }

  public class NeuronLayer : Collection<Neuron>
  {

  }

  public class Demo
  {
    static void Main(string[] args)
    {
      var neuron1 = new Neuron();
      var neuron2 = new Neuron();
      var layer1 = new NeuronLayer();
      var layer2 = new NeuronLayer();

      neuron1.ConnectTo(neuron2);
      neuron1.ConnectTo(layer1);
      layer1.ConnectTo(layer2);
    }
  }
}
